lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    , toKeyHash
+    ) where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( xpubPublicKey )
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    , hex
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..), accountPubKey )
+import Crypto.Hash.Utils
+    ( blake2b224 )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text.Encoding as T
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) =
+        compare (T.decodeUtf8 $ hex sh1) (T.decodeUtf8 $ hex sh2)
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script s@(SeqState !s1 !s2 !ixs !rpk !prefix !scripts) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        toVerKey = deriveAddressPublicKey (accountPubKey s2) MultisigScript
+        minIndex = getIndex @'Soft minBound
+        scriptAddressGap = 10
+        ourVerKeys =
+            map (\ix -> toVerKey (toEnum (fromInteger $ toInteger $ minIndex + ix)))
+            [0 .. scriptAddressGap - 1]
```

Comment:
It is actually going to be more complex than that. We can't compute "our ver keys" statically because it depends on previous events. So this must be part of the state in the form of an `AddressPool`, and updated as part of `isShared` here. The list of "our ver keys" will be actually growing from script to script, because as we discover scripts we may extend the max index (done automatically by the  'AddressPool' data-structure).

Reply:
indeed, see my last commits and property tests that illustrate what was I after